File: src/CSharp/Meshes/ExtractEdges.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Forms;
using System.Diagnostics;

using Kitware.VTK;

namespace ActiViz.Examples {
   public partial class Form1 : Form {
      public Form1() {
         InitializeComponent();
      }


      private void renderWindowControl1_Load(object sender, EventArgs e) {
         try {
            ExtractEdges();
         }
         catch(Exception ex) {
            MessageBox.Show(ex.Message, "Exception", MessageBoxButtons.OK);
         }
      }


      private void ExtractEdges() {
         vtkSphereSource sphereSource = vtkSphereSource.New();
         sphereSource.Update();

         Debug.WriteLine("Sphere" + Environment.NewLine + "----------");
         Debug.WriteLine("There are " + sphereSource.GetOutput().GetNumberOfCells() + " cells.");
         Debug.WriteLine("There are " + sphereSource.GetOutput().GetNumberOfPoints() + " points.");

         vtkExtractEdges extractEdges = vtkExtractEdges.New();
#if VTK_MAJOR_VERSION_5
         extractEdges.SetInputConnection(sphereSource.GetOutputPort());
#else
         extractEdges.SetInputData(sphereSource);
#endif
         extractEdges.Update();

         vtkCellArray lines = extractEdges.GetOutput().GetLines();
         vtkPoints points = extractEdges.GetOutput().GetPoints();

         Debug.WriteLine(Environment.NewLine + "Edges" + Environment.NewLine + "----------" );
         Debug.WriteLine("There are " + lines.GetNumberOfCells() + " cells." );
         Debug.WriteLine("There are " + points.GetNumberOfPoints() + " points." );

         // Traverse all of the edges
         for(int i = 0; i < extractEdges.GetOutput().GetNumberOfCells(); i++) {
            //Debug.WriteLine("Type: " + extractEdges.GetOutput().GetCell(i).GetClassName() );
            vtkLine line = vtkLine.SafeDownCast(extractEdges.GetOutput().GetCell(i));
            Debug.WriteLine("Line " + i + " : " + line );
         }

         // Visualize the edges

         // Create a mapper and actor
         vtkPolyDataMapper mapper = vtkPolyDataMapper.New();
#if VTK_MAJOR_VERSION_5
         mapper.SetInputConnection(extractEdges.GetOutputPort());
#else
         mapper.SetInputData(extractEdges);
#endif
         vtkActor actor = vtkActor.New();
         actor.SetMapper(mapper);
         // get a reference to the renderwindow of our renderWindowControl1
         vtkRenderWindow renderWindow = renderWindowControl1.RenderWindow;
         // renderer
         vtkRenderer renderer = renderWindow.GetRenderers().GetFirstRenderer();
         // set background color
         renderer.SetBackground(1, 1, 1);
         // add our actor to the renderer
         renderer.AddActor(actor);
      }
   }
}

File: src/Cxx/Meshes/ExtractEdges.cxx
#include <vtkSmartPointer.h>
#include <vtkPolyData.h>
#include <vtkExtractEdges.h>
#include <vtkSphereSource.h>
#include <vtkCellArray.h>
#include <vtkPoints.h>
#include <vtkPolyDataMapper.h>
#include <vtkActor.h>
#include <vtkRenderWindow.h>
#include <vtkRenderer.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkLine.h>

int main(int, char *[])
{
  vtkSmartPointer<vtkSphereSource> sphereSource =
    vtkSmartPointer<vtkSphereSource>::New();
  sphereSource->Update();

  std::cout << "Sphere" << endl << "----------" << endl;
  std::cout << "There are " << sphereSource->GetOutput()->GetNumberOfCells() << " cells." << std::endl;
  std::cout << "There are " << sphereSource->GetOutput()->GetNumberOfPoints() << " points." << std::endl;

  vtkSmartPointer<vtkExtractEdges> extractEdges =
    vtkSmartPointer<vtkExtractEdges>::New();
  extractEdges->SetInputConnection(sphereSource->GetOutputPort());
  extractEdges->Update();

  vtkCellArray* lines= extractEdges->GetOutput()->GetLines();
  vtkPoints* points = extractEdges->GetOutput()->GetPoints();

  std::cout << std::endl << "Edges" << endl << "----------" << std::endl;
  std::cout << "There are " << lines->GetNumberOfCells() << " cells." << std::endl;
  std::cout << "There are " << points->GetNumberOfPoints() << " points." << std::endl;

  // Traverse all of the edges
  for(vtkIdType i = 0; i < extractEdges->GetOutput()->GetNumberOfCells(); i++)
  {
    //std::cout << "Type: " << extractEdges->GetOutput()->GetCell(i)->GetClassName() << std::endl;
    vtkSmartPointer<vtkLine> line = dynamic_cast<vtkLine*>(extractEdges->GetOutput()->GetCell(i));
    std::cout << "Line " << i << " : " << *line << std::endl;
  }

  // Visualize the edges

  // Create a mapper and actor
  vtkSmartPointer<vtkPolyDataMapper> mapper =
    vtkSmartPointer<vtkPolyDataMapper>::New();
  mapper->SetInputConnection(extractEdges->GetOutputPort());
  vtkSmartPointer<vtkActor> actor =
    vtkSmartPointer<vtkActor>::New();
  actor->SetMapper(mapper);

  // Create a renderer, render window, and interactor
  vtkSmartPointer<vtkRenderer> renderer =
    vtkSmartPointer<vtkRenderer>::New();
  vtkSmartPointer<vtkRenderWindow> renderWindow =
    vtkSmartPointer<vtkRenderWindow>::New();
  renderWindow->AddRenderer(renderer);
  vtkSmartPointer<vtkRenderWindowInteractor> renderWindowInteractor =
    vtkSmartPointer<vtkRenderWindowInteractor>::New();
  renderWindowInteractor->SetRenderWindow(renderWindow);

  // Add the actor to the scene
  renderer->AddActor(actor);
  renderer->SetBackground(1,1,1); // Background color white

  // Render and interact
  renderWindow->Render();
  renderWindowInteractor->Start();

  return EXIT_SUCCESS;
}
